server/src/main/java/io/crate/expression/tablefunctions/GenerateSubscripts.java

```diff
@@ -0,0 +1,169 @@
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.expression.tablefunctions;
+
+import io.crate.data.Input;
+import io.crate.data.Row;
+import io.crate.metadata.FunctionName;
+import io.crate.metadata.TransactionContext;
+import io.crate.metadata.functions.Signature;
+import io.crate.metadata.pgcatalog.PgCatalogSchemaInfo;
+import io.crate.metadata.tablefunctions.TableFunctionImplementation;
+import io.crate.types.DataTypes;
+import io.crate.types.RowType;
+
+import java.util.List;
+import java.util.Locale;
+
+import static io.crate.metadata.functions.TypeVariableConstraint.typeVariable;
+import static io.crate.types.DataTypes.INTEGER;
+import static io.crate.types.TypeSignature.parseTypeSignature;
+
+public final class GenerateSubscripts<T> extends TableFunctionImplementation<T> {
+
+    public static final FunctionName NAME = new FunctionName(PgCatalogSchemaInfo.NAME, "generate_subscripts");
+    private static final RowType RETURN_TYPE = new RowType(List.of(INTEGER), List.of(NAME.name()));
+
+
+    public static void register(TableFunctionModule module) {
+        module.register(
+            Signature.table(
+                NAME,
+                parseTypeSignature("array(E)"),
+                DataTypes.INTEGER.getTypeSignature(),
+                DataTypes.INTEGER.getTypeSignature()
+            ).withTypeVariableConstraints(typeVariable("E")),
+            GenerateSubscripts::new
+        );
+        module.register(
+            Signature.table(
+                NAME,
+                parseTypeSignature("array(E)"),
+                DataTypes.INTEGER.getTypeSignature(),
+                DataTypes.BOOLEAN.getTypeSignature(),
+                DataTypes.INTEGER.getTypeSignature()
+            ).withTypeVariableConstraints(typeVariable("E")),
+            GenerateSubscripts::new
+        );
+    }
+
+    private final Signature signature;
+    private final Signature boundSignature;
+
+    private GenerateSubscripts(Signature signature, Signature boundSignature) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+    }
+
+    private static int getNumRows(List<?> array, int depthLevel) {
```

Comment:
lgtm

it is a bit hard to follow sometime some var names and messages with them, we have dim, level, target level, depth level, etc. but i am not sure how it can be improved.